sw/device/lib/dif/dif_otp_ctrl.h

```diff
@@ -0,0 +1,773 @@
+ * hardware-managed digests. An error is returned for SW partitions and the
+ * lifecycle state partition.
+ *
+ * @param otp An OTP handle.
+ * @param partition The partition to digest and lock.
+ * @return The result of the operation.
+ */
+// Open Q: do we want a SW version, as a convenience for programming the digest
+// of a SW partition?
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_dai_result_t dif_otp_ctrl_dai_hw_digest(
+    const dif_otp_ctrl_t *otp, dif_otp_ctrl_partition_t partition);
+
+/**
+ * Gets the buffered digest value for the given partition.
+ *
+ * Note that this value is only updated when the device is reset. A value of
+ * zero indicates that the digest has not been burned in yet (and, therefore,
+ * that the partition is writeable).
+ *
+ * The lifecycle partition does not have a digest and will result in an error
+ * being returned.
+ *
+ * @param otp An OTP handle.
+ * @param partition The parittion to get a digest for.
+ * @param[out] digest Out-param for the digest.
+ * @return The result of the operation.
+ */
+// Open Q: should a value of zero be returned as an error, indicating that the
+// digest doesn't exist yet?
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_get_digest(
+    const dif_otp_ctrl_t *otp, dif_otp_ctrl_partition_t partition,
+    uint64_t *digest);
+
+/**
+ * Gets the buffered lifecycle state.
+ *
+ * TODO(mcyoung): More docs when we decide on the interface for this function.
+ */
+// Open Q: do we want to wrap `state` in a struct? Right now it just gets
+// decayed into a pointer so the length parameter is mostly a lie.
+//
+// Alternatively, should this return a bit vector indicating which lifecycle
+// state half-words have been burned?
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_result_t dif_otp_ctrl_get_lifecycle_state(
+    const dif_otp_ctrl_t *otp, uint16_t state[12]);
+
+/**
+ * Performs a memory-mapped read of the given partition, if it supports them.
+ *
+ * In particular, this function will read `len` words, starting at `address`,
+ * relative to the start of `partition`.
+ *
+ * The same caveats for `dif_otp_ctrl_dai_read_start()` apply to `address`; in
+ * addition, `address + len` must also be in-range and must not overflow.
+ *
+ * @param otp An OTP handle.
+ * @param partition The partition to read from.
+ * @param address A partition-releative address to read from.
+ * @param[out] buf A buffer of words to write read values to.
+ * @param len The number of words to read.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_otp_ctrl_dai_result_t dif_otp_ctrl_read_polled(
+    const dif_otp_ctrl_t *otp, dif_otp_ctrl_partition_t partition,
+    uint32_t address, uint32_t *buf, size_t len);
+
+// Open Q: do we want a function for doing reads and writes to the TEST mmap
+// region? It doesn't seem like the best idea, seeing as that region is
+// basically a "send magic commands to the closed-source macro" that feels
+// beyond the scope of the DIF library.
```

Comment:
Yes, there should be a function that copies data directly into the `TEST` region. If the test partition is RW then there should be a way of reading data back as well.